arangod/RestHandler/RestRepairHandler.cpp

```diff
@@ -0,0 +1,699 @@
+
+    VPackSlice planCollections = plan.get("Collections");
+
+    ResultT<VPackBufferPtr> healthResult = getFromAgency("Supervision/Health");
+
+    if (healthResult.fail()) {
+      LOG_TOPIC(ERR, arangodb::Logger::CLUSTER)
+          << "RestRepairHandler::repairDistributeShardsLike: "
+          << "Failed to fetch server health result";
+      generateError(rest::ResponseCode::SERVER_ERROR,
+                    healthResult.errorNumber(), healthResult.errorMessage());
+
+      return RestStatus::FAIL;
+    }
+
+    VPackSlice supervisionHealth(healthResult.get()->data());
+
+    ResultT<std::map<CollectionID, ResultT<std::list<RepairOperation>>>>
+        repairOperationsByCollectionResult =
+            DistributeShardsLikeRepairer::repairDistributeShardsLike(
+                planCollections, supervisionHealth);
+
+    if (repairOperationsByCollectionResult.fail()) {
+      LOG_TOPIC(ERR, arangodb::Logger::CLUSTER)
+          << "RestRepairHandler::repairDistributeShardsLike: "
+          << "Error during preprocessing: "
+          << "[" << repairOperationsByCollectionResult.errorNumber() << "] "
+          << repairOperationsByCollectionResult.errorMessage();
+      generateError(rest::ResponseCode::SERVER_ERROR,
+                    repairOperationsByCollectionResult.errorNumber(),
+                    repairOperationsByCollectionResult.errorMessage());
+
+      return RestStatus::FAIL;
+    }
+    std::map<CollectionID, ResultT<std::list<RepairOperation>>>&
+        repairOperationsByCollection = repairOperationsByCollectionResult.get();
+
+    VPackBuilder response;
+    response.add(VPackValue(VPackValueType::Object));
+
+    bool errorOccurred;
+
+    if (repairOperationsByCollection.empty()) {
+      response.add("message", VPackValue("Nothing to do."));
+      errorOccurred = false;
+    } else {
+      std::stringstream message;
+      message << "Repairing " << repairOperationsByCollection.size()
+              << " collections";
+
+      response.add("collections", VPackValue(VPackValueType::Object));
+
+      bool allCollectionsSucceeded = repairAllCollections(
+          planCollections, repairOperationsByCollection, response);
+
+      if (!allCollectionsSucceeded) {
+        responseCode = rest::ResponseCode::SERVER_ERROR;
+      }
+
+      errorOccurred = !allCollectionsSucceeded;
+
+      response.close();
+    }
+
+    response.close();
+
+    generateResult(responseCode, response, errorOccurred);
+
+    clusterInfo->loadPlan();
+
+    return RestStatus::DONE;
+  } catch (Exception& e) {
+    LOG_TOPIC(ERR, arangodb::Logger::CLUSTER)
+        << "RestRepairHandler::repairDistributeShardsLike: "
+        << "Caught exception: " << e.message();
+    generateError(rest::ResponseCode::SERVER_ERROR, e.code());
+
+    if (ClusterInfo* clusterInfo = ClusterInfo::instance()) {
+      clusterInfo->loadPlan();
+    }
+
+    return RestStatus::FAIL;
+  }
+}
+
+bool RestRepairHandler::repairAllCollections(
```

Comment:
Because of timing issues i interrupted the Review here and will continue later.

Reply:
I've adressed the requested changes locally; I'll push as soon as you've completed your review.